### PR TITLE
--persist-replace argument

### PR DESCRIPTION
**Pull request by tonykploomber**
## Describe your changes

Add `--persist-replace` argument

## Issue number

Closes https://github.com/ploomber/jupysql/issues/154

## Checklist before requesting a review

- [ ] Performed a self-review of my code
- [ ] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [ ] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [ ] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--440.org.readthedocs.build/en/440/

### CODE REVIEW

File: src/tests/test_magic.py
Comment (yafimvo) [inline src/tests/test_magic.py:230]: Are we testing the initial case when the user uses `--persist` after `--persist`?

```
ValueError: Table 'plot_data' already exists.
If you need help solving this issue, send us a message: https://ploomber.io/community
```

If no, I think we should add it.

File: src/tests/test_magic.py
Comment (yafimvo) [inline src/tests/test_magic.py:183]: To make sure we are not missing anything, can we also test on different tables (i.e. `number_table`) and on their actual values (and not just the count)? Also, I think it will be easier if we'll use parameters here.

File: src/tests/test_magic.py
Comment (yafimvo) [reply src/tests/test_magic.py:230]: And another 2 cases I thought of:
1. When we are using `--persist-replace` at the beginning (without `--persist`)
2. When we are using `--persist-replace` twice

**Review by yafimvo — CHANGES_REQUESTED** [review]
added some comments.

Question - can we detect if the user tries to use `--persist` for the 2nd time and suggest him to use `--persist-replace` instead?

File: src/tests/test_magic.py
Comment (yafimvo) [reply src/tests/test_magic.py:183]: > To make sure we are not missing anything, can we also test on different tables (i.e. `number_table`) and on their actual values (and not just the count)? Also, I think it will be easier if we'll use parameters here.

this comment relates to all upcoming 4 tests

File: src/sql/magic.py
Comment (tonykploomber) [reply src/sql/magic.py:356]: Good catch!

File: src/sql/magic.py
Comment (edublancas) [reply src/sql/magic.py:184]: the convention is to use one character for short options and we're following that convention for all new options, I think -P is better

File: src/sql/magic.py
Comment (edublancas) [reply src/sql/magic.py:354]: the PR with our own types hasn't been merged: https://github.com/ploomber/jupysql/pull/431

Comment (tonykploomber) [reply src/sql/magic.py:354]: I can modify that after #431 got merged

File: src/sql/magic.py
Comment (tonykploomber) [reply src/sql/magic.py:184]: Good point, let's use `-P` for now

File: src/sql/magic.py
Comment (edublancas) [reply src/sql/magic.py:354]: #431 has been merged

Comment (tonykploomber) [reply src/sql/magic.py:355]: Cool, now the exception trace is cleaner

Screenshot 2023-04-26 at 12 16 07 AM

File: src/tests/test_magic.py
Comment (tonykploomber) [reply src/tests/test_magic.py]: I added [def query_and_save_as_dataframe](https://github.com/ploomber/jupysql/pull/440/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R198-R202) to reuse

File: src/tests/test_magic.py
Comment (tonykploomber) [reply src/tests/test_magic.py:204]: 1. Added different tables as parameters
2. Added more tests 
(When we are using --persist-replace at the beginning (without --persist)
When we are using --persist-replace twice)

**Comment by tonykploomber** [comment]
@yafimvo 

I checked the [preview doc](https://github.com/ploomber/jupysql/pull/440/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R198-R202), somehow there is the empty cell in the original --persist section, do you think our PR affect this?

Screenshot 2023-04-26 at 12 19 48 AM

**Comment by tonykploomber** [comment]
@idomic 

Now if we use --persist twice it will suggest to use `--persist-replace` 

Screenshot 2023-04-26 at 12 16 07 AM

File: src/sql/magic.py
Comment (yafimvo) [reply src/sql/magic.py:355]: Shouldn't we raise here `exceptions.ValueError` instead of `ValueError`?

File: src/tests/test_magic.py
Comment (yafimvo) [inline src/tests/test_magic.py:225]: can we remove this comment? Not like before, I'm not sure the length is something we should specifically mention here.

File: src/tests/test_magic.py
Comment (yafimvo) [inline src/tests/test_magic.py:253]: Same here

File: src/tests/test_magic.py
Comment (yafimvo) [inline src/tests/test_magic.py:281]: Same here. I don't think the lengths of the saved tables shouldn't matter since we are comparing with the expected_results

File: src/tests/test_magic.py
Comment (yafimvo) [inline src/tests/test_magic.py:298]: Same here

File: src/tests/test_magic.py
Comment (yafimvo) [inline src/tests/test_magic.py:317]: Let's remove this line and check the name of the error as well (which should be "ValueError")

File: src/tests/test_magic.py
Comment (yafimvo) [inline src/tests/test_magic.py:313]: How are we testing `--persist` and `--persist-replace` here? If we are not, let's remove this line. If we are, let's change the test name which implies something different (`--persist` and `--append`)

File: src/tests/test_magic.py
Comment (yafimvo) [inline src/tests/test_magic.py:342]: Let's remove this line since the test name is descriptive enough

**Review by yafimvo — CHANGES_REQUESTED** [review]
Added some minor comments

File: src/tests/test_magic.py
Comment (yafimvo) [inline src/tests/test_magic.py:198]: I think the name might confuse since we are running a specific query on a given table (`SELECT * FROM table`).
Maybe we can change it to something like `get_table_rows_as_dataframe` (or a better one) and initiate `limit` with a default number, so the purpose of this function would be to "return X rows from the given table as dataframe."

Question - I see you are mostly fetching only 1 line from the table. Does it matter for the tests? If not, let's put a default number of 5, if it matters maybe the default number should be 1.

this way, we can turn this line:

```
saved_df_name = query_and_save_as_dataframe(ip, table_name=test_table, limit=1)
```

to 

```
saved_df_name = get_table_rows_as_dataframe(ip, table_name=test_table)
```

which might be easier to read, use and maintain.

what do you think?

File: src/tests/test_magic.py
Comment (tonykploomber) [reply src/tests/test_magic.py:313]: Looks like we didn't cover the test case that `--persist and --persist-replace` used together.

Do you think we should make it pass or just fail it if it's happening?

Comment (yafimvo) [reply src/tests/test_magic.py:313]: I asked this because that is what is written in the comment which is different than the actual test and the test name.
As for your question, since you [implemented this behaviour](https://github.com/ploomber/jupysql/pull/440/files/f53e7a2765268154bb0f4dc4d6efac688641fb1e#diff-527caf90dd08297f5253e826bd98fea0b6f4fbd44fe9f20472976f77868dac06R353) - we should check this case and expect for this value error `You cannot both replace and append to a dataframe at same time, please only use either of them`.

Comment (yafimvo) [reply src/tests/test_magic.py:311]: As for your question, I think it's weird behavior but I wouldn't fail it. 
Maybe add a warning and ignore the `--persist`? @idomic @edublancas what do you think?

File: src/tests/test_magic.py
Comment (yafimvo) [reply src/tests/test_magic.py:198]: yes, but do we actually test these different lengths? I think in most of the cases I've seen, we test the actual values, so the limit doesn't make much difference.

**Comment by tonykploomber** [comment]
Still figuring out why the empty cell was generated...

**Comment by yafimvo** [comment]
> checked the [preview doc](https://github.com/ploomber/jupysql/pull/440/files#diff-d327143929530bcece2cae65d59465708cdf87f499fd83b4bd939c332150f713R198-R202), somehow there is the empty cell in the original --persist section, do you think our PR affect this?

I don't think it's related to your changes in this PR, but it might be a good idea to remove it since it's a small and simple change that will make the documentation cleaner without affecting anything else.

File: src/tests/test_magic.py
Comment (tonykploomber) [reply src/tests/test_magic.py:198]: Oh sorry I think we did not test the length, but the actual value

I can rename the function to `get_table_rows_as_dataframe` and remove the limit part, although we will always use same dataframe to store by `--persist` and -`-persist-replace`

File: src/tests/test_magic.py
Comment (tonykploomber) [reply src/tests/test_magic.py]: Fix in https://github.com/ploomber/jupysql/pull/440/commits/b99cf0952ce2d429e93815e2ea966ce52218b55c

**Comment by tonykploomber** [comment]
Please review again! 
@idomic @yafimvo

File: src/sql/magic.py
Comment (tonykploomber) [reply src/sql/magic.py:360]: This is the test case when `--persist` and `--persist-replace` got executed together

https://github.com/ploomber/jupysql/pull/440#discussion_r1179231581

I thought we gonna ignore + adding warning message

File: src/sql/magic.py
Comment (tonykploomber) [reply src/sql/magic.py]: Addressed!

Comment (tonykploomber) [reply src/sql/magic.py]: [d56695d](https://github.com/ploomber/jupysql/pull/440/commits/d56695d7c8bc48734b439394ac01ff4ac493ee8b)

File: doc/integrations/pandas.md
Comment (tonykploomber) [reply doc/integrations/pandas.md]: Gotcha, addressed

File: src/sql/magic.py
Comment (tonykploomber) [reply src/sql/magic.py]: Addressed, [5ed823f](https://github.com/ploomber/jupysql/pull/440/commits/5ed823ff8405c2e7ded90bd102ba1c0c1add5095)

File: src/sql/magic.py
Comment (tonykploomber) [reply src/sql/magic.py:360]: @idomic Let's check again!

Comment (tonykploomber) [reply src/sql/magic.py:360]: I understand wrongly earlier indeed, fix in [13a3a5b](https://github.com/ploomber/jupysql/pull/440/commits/13a3a5bfd621daa4c363e9ea1714b41c9551c4b4)

File: src/tests/test_magic.py
Comment (yafimvo) [inline src/tests/test_magic.py:287]: How do we know that the values changed? I think we should add another table to make sure the values are changed.

File: src/tests/test_magic.py
Comment (yafimvo) [inline src/tests/test_magic.py:351]: Same as before, we need another table (different) to make sure the relevant data is used.

File: src/tests/test_magic.py
Comment (yafimvo) [inline src/tests/test_magic.py:379]: Please use another table here as well to make sure `--persist-replace` uses different data

**Review by yafimvo — CHANGES_REQUESTED** [review]
I think we're missing the most important part of the test. We need to make sure `--persist-replace` overrides the previous data. To do it, we need to use 2 different datasets.


```
saved_df_name = get_table_rows_as_dataframe(ip, table_name=test_table)
ip.run_cell(f"%sql --persist sqlite:// {saved_df_name}")
saved_df_name = get_table_rows_as_dataframe(ip, table_name=test_table)
ip.run_cell(f"%sql --persist-replace sqlite:// {saved_df_name}")
assert out.result == expected_result
```

even if `--persist-replace` doesn't work, it will return True.
Please note, this is valid for the contrapositive tests as well (i.e. `test_persist_replace_no_override`). If you are using the same data, it will be difficult to determine whether it was not overridden.

**Comment by tonykploomber** [comment]
@yafimvo 
Actually we've discussed this about this is [previous comment](https://github.com/ploomber/jupysql/pull/440#discussion_r1177544490). That's why previous `query_and_save_as_dataframe ` function has limit parameter to generate a different dataset to test the override effect.

We can also just get the different dataframe by putting different `table_name` to `get_table_rows_as_dataframe`.

Handling

File: src/tests/test_magic.py
Comment (tonykploomber) [reply src/tests/test_magic.py:399]: Fix, please review again!

[2d4a54c](https://github.com/ploomber/jupysql/pull/440/commits/2d4a54cf5deb497aa3cbf5b4a982aa2c47ca2f98)

File: src/tests/test_magic.py
Comment (yafimvo) [inline src/tests/test_magic.py:199]: Let's add a short description to this function.
I think we can rename `table_name` -> `table` and `desire_saved_df_name` -> `name`
I also think we can add an optional parameter `limit = 1`.

Now, this function returns a single row from a specified table as a dataframe, with the dataframe named after the table. Then, you can change the number of rows and the name of the dataframe.

File: src/tests/test_magic.py
Comment (yafimvo) [inline src/tests/test_magic.py:280]: To make it more readable, let's use the returned value

```
table_df = get_table_rows_as_dataframe(ip, table)
```

(this is relevant for other places as well)

**Review by yafimvo — CHANGES_REQUESTED** [review]
> Actually we've discussed this about this is [previous comment](https://github.com/ploomber/jupysql/pull/440#discussion_r1177544490). That's why previous `query_and_save_as_dataframe ` function has limit parameter to generate a different dataset to test the override effect.

Since we didn't check for different lengths, I suggested setting the limit parameter as optional and making this function return X rows of a given table as a dataframe.

> We can also just get the different dataframe by putting different `table_name` to `get_table_rows_as_dataframe`.

I see you implemented something like this.

LGTM. just minor fixes.

**Comment by tonykploomber** [comment]
This is blocked by [CI test failing issue](https://github.com/ploomber/jupysql/issues/495)

File: CHANGELOG.md
Comment (edublancas) [inline CHANGELOG.md:9]: clearer changelog entry:

```suggestion
* [Feature] Adds `--persist-replace` argument to replace existing tables when persisting data frames (#440)
```

File: src/sql/magic.py
Comment (edublancas) [inline src/sql/magic.py:381]: what happens if we pass append and persist_replace? sounds like it should also fail, right? so it sounds like if more than one option is set in {persist, append, persist_replace} we should throw an error

File: src/sql/magic.py
Comment (edublancas) [inline src/sql/magic.py:382]: this should raise a UsageError - please add a unit test as well

File: src/sql/magic.py
Comment (edublancas) [inline src/sql/magic.py:533]:
```suggestion
        except ValueError:
            raise exceptions.ValueError(
                f"Table {table_name!r} already exists. Consider using --persist-replace to drop the table before persisting the data frame"
```

File: src/sql/magic.py
Comment (edublancas) [inline src/sql/magic.py:377]: this should be a usageerror

**Review by edublancas — CHANGES_REQUESTED** [review]
it looks good overall, added some minor comments

File: src/sql/magic.py
Comment (tonykploomber) [reply src/sql/magic.py:381]: Yes, it will fail
Screenshot 2023-05-24 at 3 47 41 PM

File: src/sql/magic.py
Comment (tonykploomber) [reply src/sql/magic.py:382]: [In earlier discussion] (https://github.com/ploomber/jupysql/pull/440#discussion_r1177515061) they thought we can still apply the persist_replace and throw the warning

What do you think? Should we just raise the UsageError with `Please use either --persist or --persist-replace` ?

File: src/sql/magic.py
Comment (edublancas) [reply src/sql/magic.py:381]: alright, this makes sense

File: src/sql/magic.py
Comment (tonykploomber) [reply src/sql/magic.py]: Fix!

**Comment by tonykploomber** [comment]
Please review again

File: CHANGELOG.md
Comment (edublancas) [inline CHANGELOG.md]: looks like some rebase went bad? I see some entries in the changelog are deleted

**Review by edublancas — CHANGES_REQUESTED** [review]

File: CHANGELOG.md
Comment (tonykploomber) [reply CHANGELOG.md]: Can you check again? I rebase again
Screenshot 2023-05-27 at 4 31 35 AM

**Review by edublancas — APPROVED** [review]

**Pull request merged**
